File: vuln-sandbox/python/taint/xss-2hop.py
# Taint: XSS — 2-hop (Python)
# Source: request.args.get('msg') → intermediate: build_page() → Sink: Markup
# Expected: TruePositive (taint/cwe79)
from flask import Flask, request
from markupsafe import Markup

app = Flask(__name__)

def build_page(msg):
    return Markup(f"<div>{msg}</div>")         # SINK: unescaped HTML markup

@app.route('/message')
def message():
    msg = request.args.get('msg')              # SOURCE: HTTP request param
    return build_page(msg)                     # INTERMEDIATE

File: vuln-sandbox/node/taint/xss-2hop.js
// Taint: XSS — 2-hop (JS)
// Source: req.query.msg → intermediate: renderMessage() → Sink: innerHTML
// Expected: TruePositive (taint/cwe79)
const express = require('express');
const app = express();

function renderMessage(msg) {
    document.getElementById('output').innerHTML = msg; // SINK: innerHTML
}

app.get('/message', (req, res) => {
    const msg = req.query.msg;                 // SOURCE: HTTP request param
    renderMessage(msg);                        // INTERMEDIATE
    res.send('rendered');
});
